### PR TITLE
Automatically build libstd

### PR DESCRIPTION
**Pull request by RalfJung**
Cc https://github.com/solson/miri/issues/520

### CODE REVIEW

File: src/bin/cargo-miri.rs [hunk 1]
```diff
@@ -34,6 +48,145 @@ fn show_version() {
...
+    print!("{} [Y/n] ", question);
+    io::stdout().flush().unwrap();
+    io::stdin().read_line(&mut buf).unwrap();
+    let answer = match buf.trim().to_lowercase().as_ref() {
+        "" | "y" | "yes" => true,
+        "n" | "no" => false,
+        a => show_error(format!("I do not understand `{}`", a))
+    };
+    if !answer {
+        show_error(format!("Aborting as per your request"))
```
Comment (bjorn3) [inline src/bin/cargo-miri.rs]: could you move this to the match case returning false?

Comment (RalfJung) [reply src/bin/cargo-miri.rs]: Hm, I found it nicer to first convert-to-bool and then act on the bool. But I do not particularly care.

File: src/bin/cargo-miri.rs [hunk 2]
```diff
@@ -34,6 +48,149 @@ fn show_version() {
...
+        val.starts_with("--manifest-path=")
+    });
+
+    let mut metadata = if let Ok(metadata) = cargo_metadata::metadata(
+        manifest_path_arg.as_ref().map(AsRef::as_ref),
+    )
+    {
+        metadata
+    } else {
+        let _ = std::io::stderr().write_fmt(format_args!(
```
Comment (oli-obk) [inline src/bin/cargo-miri.rs]: `eprintln!` should be equivalent, should it not?

File: src/bin/cargo-miri.rs (hunk 1 above)
Comment (oli-obk) [reply src/bin/cargo-miri.rs]: I also think that it should be in the match.

File: src/bin/cargo-miri.rs [hunk 3]
```diff
@@ -34,6 +48,149 @@ fn show_version() {
...
+        } else {
+            println!("Installing rust-src component: `rustup component add rust-src`");
+        }
+        if !Command::new("rustup").args(&["component", "add", "rust-src"]).status().unwrap().success() {
+            show_error(format!("Failed to install rust-src component"));
+        }
+    }
+
+    // Next, we need our own libstd. We will do this work in ~/.miri.
+    let dir = dirs::home_dir().unwrap().join(".miri");
```
Comment (oli-obk) [inline src/bin/cargo-miri.rs]: use https://crates.io/crates/directories

File: src/bin/cargo-miri.rs [hunk 4]
```diff
@@ -34,6 +48,149 @@ fn show_version() {
...
+    if !dir.exists() {
+        fs::create_dir(&dir).unwrap();
+    }
+    // The interesting bit: Xargo.toml
+    File::create(dir.join("Xargo.toml")).unwrap()
+        .write_all(br#"
```
Comment (oli-obk) [inline src/bin/cargo-miri.rs]: maybe put these into files in a separate directory and use `include_bytes!`?

**Comment by RalfJung** [comment]
> I'm not very up to date on the cli working group things, but I think they are at a point where it might make sense to look at the best practices and crates for writing cli tools. Not a blocker for this PR, feel free to punt to an issue and assign to me

Done.

File: src/bin/cargo-miri.rs (hunk 2 above)
Comment (RalfJung) [reply src/bin/cargo-miri.rs]: Probably. This code is *old*, I just moved it around.^^

File: src/bin/cargo-miri.rs (hunk 3 above)
Comment (RalfJung) [reply src/bin/cargo-miri.rs]: Sure, whatever. libstd recommends `dirs`.

File: src/bin/cargo-miri.rs (hunk 4 above)
Comment (RalfJung) [reply src/bin/cargo-miri.rs]: You think that's worth it? They are rather short.

File: src/bin/cargo-miri.rs [hunk 5]
```diff
@@ -34,6 +48,143 @@ fn show_version() {
...
+        } else {
+            println!("Installing rust-src component: `rustup component add rust-src`");
+        }
+        if !Command::new("rustup").args(&["component", "add", "rust-src"]).status().unwrap().success() {
+            show_error(format!("Failed to install rust-src component"));
+        }
+    }
+
+    // Next, we need our own libstd. We will do this work in ~/.miri.
+    let dir = directories::UserDirs::new().unwrap().home_dir().join(".miri");
```
Comment (oli-obk) [inline src/bin/cargo-miri.rs]: not quite what I had in mind.

```rust
// on linux: `.cache/mirì`
directories::ProjectDirs::from("miri", "miri", "miri").cache_dir()
```

Comment (RalfJung) [reply src/bin/cargo-miri.rs]: Oh, fancy. I like it :D

**Comment by RalfJung** [comment]
Uh... which directory will this be on Windows? We need to know for CI.^^

**Comment by RalfJung** [comment]
Uh... now the path is different on Linux and macOS, and travis needs to know that. :/

**Pull request merged**